app/models/event_stream.rb

```diff
@@ -59,57 +60,75 @@ def emit_notifications
 
   # TODO: Consider moving since this is EmsEvent specific. group, group_level and group_name exposed as a virtual columns for reports/api.
   def self.event_groups
-    core_event_groups = ::Settings.event_handling.event_groups.to_hash
-    Settings.ems.each_with_object(core_event_groups) do |(_provider_type, provider_settings), event_groups|
-      provider_event_groups = provider_settings.fetch_path(:event_handling, :event_groups)
-      next unless provider_event_groups
-      DeepMerge.deep_merge!(
-        provider_event_groups.to_hash, event_groups,
-        :preserve_unmergeables => false,
-        :overwrite_arrays      => false
-      )
+    @event_groups ||= begin # TODO: this cache needs to be busted whenever settings are changed (and also in test)
+      core_event_groups = ::Settings.event_handling.event_groups.to_hash
+      Settings.ems.each_with_object(core_event_groups) do |(_provider_type, provider_settings), event_groups|
+        provider_event_groups = provider_settings.fetch_path(:event_handling, :event_groups)
+        next unless provider_event_groups
+
+        DeepMerge.deep_merge!(
+          provider_event_groups.to_hash, event_groups,
+          :preserve_unmergeables => false,
+          :overwrite_arrays      => false
+        )
+      end
     end
   end
 
-  # TODO: Consider moving since this is EmsEvent specific. group, group_level and group_name exposed as a virtual columns for reports/api.
-  def self.group_and_level(event_type)
-    level = DEFAULT_GROUP_LEVEL # the level is detail as default
-    egroups = event_groups
+  private_class_method def self.partition_group_and_level_by_event_type
+    # TODO: this cache needs to be busted whenever settings are changed (and also in test)
```

Comment:
think this is fixed

Reply:
Yeah you're right - those TODO's can be removed.

Reply:
https://github.com/ManageIQ/manageiq/pull/22350